FEATURE: V2 NonFungibleToken Standard

joshuahannan opened the pull request:
Adds the v2 fungible token standard, as described in this flip: https://github.com/onflow/flips/pull/56

A few highlights:

* Turns NonFungibleToken into a contract instead of a contract interface
* Introduces a new contract interface (`NonFungibleTokenInterface`) with event specifications, functions to get the NFT and collection types that the contract implements, and functions to get important metadata views for the types in the contract. This will help external parties learn about which types a contract implements since contracts can implement multiple collection and NFT types now.
* Adds a Transferable interface to include a transfer function
* Adds getAcceptedTypes() to the Receiver interface so a receiver interface can indicate which NFT types it accepts.
* Moves standard paths inside the collection object definition because they are specific to the actual collection instead of the contract now.
* Adds type parameters to the all the events since contracts can have multiple nft types defined
* Adds a type parameter to the `createEmptyCollection` method

The contract currently has a couple errors that should be able to be resolved easily, such as it not working properly with the `createEmptyCollectionFunction` field of the `NFTCollectionData` metadata view, and the `ownedNFTs` field not conforming to the resource interface's field, but there are no other errors.
I am looking for feedback on the initial proposal. Tests will come later
There are probably more things that I am missing, but most of the details are in the forum post, so that is another good resource.

bjartek commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,404 @@
+                case Type<MetadataViews.Traits>():
+                    // exclude mintedTime and foo to show other uses of Traits
+                    let excludedTraits = ["mintedTime", "foo"]
+                    let traitsView = MetadataViews.dictToTraits(dict: self.metadata, excludedNames: excludedTraits)
```
I see that you do include other traits further down and that is a good thing!

joshuahannan replied:
I only commented the metadata views directly from the other contract here, so please make any comments about those in issues in this repo so we can keep the conversation focused on the new standard here. I can integrate any updates that we make there into this contract.

joshuahannan commented on contracts/ExampleNFT-v2-ContractInterface.cdc:
```diff
@@ -0,0 +1,51 @@
+    ///
+    /// It indicates the owner of the collection that it was deposited to.
+    ///
+    pub event Deposit(id: UInt64, to: Address?, type: Type)
```
Do you mean the storage path where the collection is stored? or the public receiver path? Those are the only paths that we can access from within the collection object, and I don't think either would be 100% accurate, because there are cases where they are stored in different places and don't have a way to pass that path information into the deposit function

bjartek replied:
Gotcha.

bjartek replied:
Can we include UUID as a field, since it is unique across all resources and we should access to it here?

joshuahannan commented on contracts/ExampleNFT-v2-ContractInterface.cdc:
```diff
@@ -0,0 +1,51 @@
+import NonFungibleToken from "./NonFungibleToken-v2.cdc"
+import MetadataViews from "./MetadataViews.cdc"
+
+pub contract interface NonFungibleTokenInterface {
```
The `Collection` interface in `NonFungibleToken` requires those

joshuahannan commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,404 @@
+        }
+
+        /// getIDs returns an array of the IDs that are in the collection
+        pub fun getIDs(): [UInt64] {
```
that would be nice to have! Once it is built, I can integrate it into this

joshuahannan commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,404 @@
+
+        /// borrowNFT gets a reference to an NFT in the collection
+        /// so that the caller can read its metadata and call its methods
+        pub fun borrowNFT(id: UInt64): &ExampleNFT.NFT{NonFungibleToken.NFT} {
```
yes, it definitely should return an optional. Good catch!

turbolent commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,390 @@
+
+    /// Return the types that the contract defines
+    pub fun getNFTTypes(): [Type] {
+        let types: [Type] = []
+        types[0] = Type<@ExampleNFT.NFT>()
+        return types
```
Any reason for not simply returning a literal?
```suggestion
        return [
            Type<@ExampleNFT.NFT>()
        ]
```

bluesign commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+    /// Interface that an account would commonly 
+    /// publish for their collection
+    pub resource interface CollectionPublic { //: MetadataViews.ResolverCollection {
+        pub fun deposit(token: @AnyResource{NFT})
```
Should we either remove `deposit` or add `getAcceptedTypes` ?

joshuahannan replied:
yes, I forgot to add `getAcceptedTypes`

bluesign commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Transferable {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun transfer(id: UInt64, receiver: Capability<&AnyResource{Receiver}>): @AnyResource{NFT}
+    }
```
Maybe we can also add a method for inbound transfer ? Also why return value `@AnyResource{NFT}` here ?

bluesign replied:
Btw wasn't this `address` before? why we changed to capability there ?

PS: I don't remember, even I could be the one suggesting capability, but address seems better choice.

bluesign replied:
this will also probably need a post check, if id in target receiver afterwards

joshuahannan replied:
I was messing around with some return type stuff and forgot to remove this here

bjartek replied:
I would strongly prefer a capability or a reference here. If you only send in address you cannot specify to what path you want something transfered to. And that makes it not so usefull.

bluesign replied:
yeah now I remember the discussion , thanks @bjartek , totally you are right.

austinkline replied:
My only knock on this versus unfolding the capability and emitting an address and a path separately is that it can be a little harder to read the capability in an event versus flat values. Either way, maintaining _where_ the item went is very important. +1

bjartek replied:
I totally agree that knowing where it went is very important. It would be very nice if capabilities had a way to get not only their address but also their path. Or some way to later be able to reason about where you can read from and expect to find it.

bluesign commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+    pub resource interface Provider {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun withdraw(withdrawID: UInt64): @AnyResource{NFT} {
+            post {
+                result.id == withdrawID: "The ID of the withdrawn token must be the same as the requested ID"
+            }
```
I don't have evidence, but have a strong feeling this will not be enough. We normally had type guarantees when adding NFT to collection ( benefit of single type ), this feels like weak guarantee here without type

joshuahannan replied:
Good point! Do you have any suggestions for how we might enforce that?

bluesign replied:
I don't have unfortunately, but I will think on this

austinkline replied:
+1 on this, a malicious contract can pretty easily get around this unfortunately. If we did it based on uuid that would be a different story but I don't see a good way to make that happen

joshuahannan replied:
I meant to have `getID()` as a function instead of requiring the ID field. I've updated the code to use that now. Thanks for pointing it out!

bluesign replied:
`getID()` has very bad synergy with this, post condition losing all meaning.

psiemens replied:
> getID() has very bad synergy with this, post condition losing all meaning. 

I see two ways this could cause a weak post condition:

1. The post condition doesn't check the type of the NFT. A user may think they're going to get TopShot 42 but actually get TopShotPack 42. Is this what you're worried about @bluesign?
2. The `getID` function is dynamic and could return a different value each time it's called. This feels like less of a problem but still prevents the post condition from giving strong guarantee.
    - I'm actually more worried about this in the general case. I think we probably want IDs to be constant identifiers of an NFT. Collections only work under this assumption. But making the ID a function could encourage developers to dynamically generate their ID based on data that could change, which feels fragile to me.

bluesign commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+        /// as long as all the other functions are there
+
+        /// Returns the NFT types that this collection can store
+        pub fun getAcceptedTypes(): [Type]
```
what if collection accepts every type ?

austinkline replied:
It might make more sense to have a method `isAcceptedType(type: Type): Bool`

joshuahannan replied:
Good question. Maybe we can make this return an optional, and have `nil` mean that it accepts every `{NFT}` type? Or it could just return one type here that is `AnyResource{NFT}`. That is probably better

bluesign commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+
+        /// Paths for the collection
+        pub let storagePath: StoragePath
+        pub let publicPath: PublicPath
+        pub let privateProviderPath: PrivatePath
+
+        /// Dictionary to hold the NFTs in the Collection
```
Maybe we can define something like another struct here: Something like Link : Type and Path , main problem in mainnet is linking currently, so maybe some improvement there would be nice. 

something like roughly:
```[
`Link(Path:/storage/TopShotCollection, Type: Type<TS.Collection>)`,
`Link(Path:/public/TopShotCollection, Type: Type<{TS.CollectionPublic>})`,
`Link(Path:/private/TopShotCollection, Type: Type<{NFT.Provider}>)`
]
```

austinkline replied:
What's the anticipated usage for this? Might make more sense to make it a dictionary so that it's easy to access these by type. Otherwise transactions are looping over an array looking for the type they care about

joshuahannan replied:
I'm changing the `getAcceptedTypes` return type to a dictionary, but I think @bluesign was talking about how to store path and link information actually. I'm not sure what the proposal here means though.

bluesign replied:
I was suggesting instead of exposing path, maybe we can expose path + type information. So user of the contract, if setting new can now know, how to setup the account correctly.

joshuahannan replied:
Isn't that what the `MetadataViews.NFTCollectionData` view is for?

bluesign commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+
+        /// From the MetadataViews Contract
+        /// borrows a reference to get metadata views for the NFTs that the contract contains
+        pub fun borrowViewResolver(id: UInt64): &{MetadataViews.Resolver}
+
```
do we need this?

joshuahannan replied:
Yeah, it is part of the `MetadataViews.ResolverCollection` interface

bluesign commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+        pub fun deposit(token: @AnyResource{NFT})
+        pub fun borrowViewResolver(id: UInt64): &{MetadataViews.Resolver}
+        pub fun getIDs(): [UInt64]
+        pub fun borrowNFT(id: UInt64): &AnyResource{NFT}
```
post conditions here ?

joshuahannan replied:
yep, gotta make sure the id is correct. adding

bluesign commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+    /// publish for their collection
+    pub resource interface CollectionPublic { //: MetadataViews.ResolverCollection {
+        pub fun deposit(token: @AnyResource{NFT})
+        pub fun borrowViewResolver(id: UInt64): &{MetadataViews.Resolver}
```
this can be removed probably

joshuahannan replied:
Why?

austinkline replied:
I'd guess we won't need this definition if the inheritance that's commented out is taken? Or are we deprecating that interface altogether?

joshuahannan replied:
it is only commented out right now because cadence currently doesn't support interfaces inheriting from other interfaces.

joshuahannan replied:
yes, the cadence team is working on it

bluesign replied:
oh I didn't see this before sorry, I was telling to remove this totally I remember ( not related to interface ) , but now I have to dig deep in memory banks.

bluesign replied:
I think I was thinking because now we have:
https://github.com/onflow/flow-nft/blob/d39375c13fdaf23977d813d9a357c4afe08c8125/contracts/NonFungibleToken-v2.cdc#L56-57

result of borrowNFT can resolve views, we don't need resolver anymore . It was needed because we couldn't downcast ( or upcast dont know the terminology ) NFT to MetadataResolver

joshuahannan replied:
really? I thought that the view resolver for the collection was still nice to have

bluesign replied:
I maybe missing the context, @bjartek knows better, but I think it was like that.

bjartek replied:
I think keeping the same name as v1 has merrit here, since it would make it easier to reason about code using both v1 and v2 of the spec. When that is said that might not be a goal that we can achieve. 

I really want to try out different alternatives when we have default implementations and see.

psiemens replied:
> I think keeping the same name as v1 has merrit here, since it would make it easier to reason about code using both v1 and v2 of the spec. When that is said that might not be a goal that we can achieve.

But to be clear, you're talking about maintaining consistency between the V1 and V2 specs, right? I'm not sure this is important if we already have to upgrade all NFT implementations from V1 to V2.

Apart from that, does the `MetadataViews.ResolverCollection` interface continue to have value on its own?

IMO it doesn't have any real value now that the NFT interface implements `ViewResolver`. I think we should remove the `borrowViewResolver` function to keep the collection interface as small as possible.

bluesign commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+
+        /// deposit takes an NFT as an argument and adds it to the Collection
+        ///
+        pub fun deposit(token: @AnyResource{NFT})
```
post condition check with `getAcceptedTypes`

joshuahannan replied:
What kind of post condition do you mean?

austinkline replied:
@bluesign do you mean a pre check that the type of the token being deposited is part of `getAcceptedTypes`? Also not sure what post condition you'd put here

bluesign replied:
@austinkline yeah I was thinking like that, with id check not checking type on withdrawal , maybe we can prevent that hole here.

bluesign commented:
Very good and much better than the current standard. I left few minor comments.

bymi15 commented:
Hi,
I was just wondering whether this could also be considered / added: https://github.com/onflow/flow-nft/issues/123
Perhaps an optional `medias` field would be nice to support a wider variety of NFTs such as videos, music, or multiple images, etc. The current structure is limited in supporting just one thumbnail image, but some platforms/marketplaces may wish to display videos / music players or multiple images in a carousel instead.

joshuahannan commented:
@bymi15 That is a separate workstream that is unrelated to this standard update. I'll respond to your issue. Thank you!

bjartek commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+    pub resource interface Collection { //: Provider, Receiver, Transferable, CollectionPublic, MetadataViews.ResolverCollection {
+
+        /// Paths for the collection
+        pub let storagePath: StoragePath
```
Personally I think the standard paths should be derived and not specified. If it includes the address it is deployed to and the name of the NFT then it is able to reason about it automatically.

bluesign replied:
oh this would be very nice, also I suggest renaming those as `defaultStoragePath` maybe. This way it shows it is not forced but suggested.

joshuahannan replied:
@bjartek Can you share how you would derive the standard paths?

bjartek replied:
Get the identifier of the resouce type, normalize it (or allow "." in path) and use that. We do this for some things in .find

joshuahannan commented:
What do we think about adding a batch deposit, batch transfer, and batch withdraw? Is that worth adding?

austinkline commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,158 @@
+        /// Paths for the collection
+        pub let storagePath: StoragePath
+        pub let publicPath: PublicPath
+        pub let privateProviderPath: PrivatePath
```
I'm torn about this one. I actually don't like that we are encouraging the use of a single provider path here. It means that we cannot really scope the access of what's being withdrawn very easily. I get the value of it, but this feels like it'll make adequately controlling what is being withdrawn almost impossible

joshuahannan replied:
I removed the provider path requirement for now

austinkline reviewed (commented):
Didn't realize this was out! Left some comments, and I'd also like to discuss how we can make use of scoped providers to encourage protecting capabilities that can withdraw. Or maybe we can bake it into both V2 standards in some way? Here's my first attempt at scoped providers as a starting point, I'd love for us to try and incorporate it as closely as we can into the standard:

https://github.com/green-goo-dao/flow-utils/blob/main/cadence/contracts/ScopedNFTProviders.cdc

bjartek commented:
I saw in mail that austin commented on the transfer discussion that has been marked as resolved. Just a heads up that there might be new comments in resolved discussions.

joshuahannan commented:
@bjartek @bluesign @austinkline and anyone else who can give feedback. Quick question about how this should be structured. Right now, I have it so that all the resource interfaces are in a contract (`FungibleToken`) and I have a separate contract interface (`NonFungibleToken-v2`) that requires the standard events and some utility functions to return the types defined by the contract and some standard views. @alilloig made the suggestion that those could be just moved into `FungibleToken` and make it a contract interface like it currently is, then any utility functions with implementations like `verifyCollectionTypes()` like we have in there right now can just be moved into a `NonFungibleTokenUtilities` contract along with other useful things like scoped providers and such. Do you think we should go with Alvaro's suggestion, or stick with what we have now and put those utility functions/scoped providers directly in the token standard contract?

alilloig commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,166 @@
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Transferable {
```
Nothing really important, but how would it sound to a native speaker to call this "Transfuser"? I believe it will be more accurate since the collection which will be implementing this is not a transferable resource by itself, but rather it has the ability to transfer stuff

austinkline replied:
Personally, I had to do a double-take on the word `Transfuser`, I've never seen it before. I get the point, though. `Transferable` suggests that the Collection is what's transferred but that isn't accurate. 

You could also think about options like:
- Sender
- Mover
- Dispatcher

I don't mind `Transferable` either for what it's worth

alilloig replied:
I actually got the word from google translate, translating a Spanish word that I believe is some device used for blood transfusions ^^u

joshuahannan replied:
I think transferable is fine for now

alilloig commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,365 @@
+        }
+
+        /// Returns the NFT types that this collection can store
+        pub fun getAcceptedTypes(): {Type: Bool} {
```
```suggestion
        pub fun getAcceptedTypes(): [Type] {
```

alilloig replied:
Okey not really, `Receiver` and `CollectionPublic` both have a `getAcceptedTypes()` method, are both needed? Probably just returning `[Type]` will do the trick right?

austinkline replied:
We'd discussed this briefly somewhere before

If we make this a list that means we are basically forcing people to loop over an array to see if something is supported. Using a dictionary makes that a non-issue, I would argue that the small bloat in data is worth it, especially since you can just get the keys of the response if an array is really what you want. This convention also lets you use the dictionary to mark what **isn't** supported which is some nice flexibility.

alilloig replied:
Thanks @austinkline !!

alilloig commented:
I don't really have any important comments, I love the proposal!! Can't wait to see what people can create using multiple NFTs and FTs on a single contract 🎉

austinkline commented:
> @bjartek @bluesign @austinkline and anyone else who can give feedback. Quick question about how this should be structured. Right now, I have it so that all the resource interfaces are in a contract (`FungibleToken`) and I have a separate contract interface (`NonFungibleToken-v2`) that requires the standard events and some utility functions to return the types defined by the contract and some standard views. @alilloig made the suggestion that those could be just moved into `FungibleToken` and make it a contract interface like it currently is, then any utility functions with implementations like `verifyCollectionTypes()` like we have in there right now can just be moved into a `NonFungibleTokenUtilities` contract along with other useful things like scoped providers and such. Do you think we should go with Alvaro's suggestion, or stick with what we have now and put those utility functions/scoped providers directly in the token standard contract?

Just to make sure I understand the question fully, is this mostly around separating the interface from any implementation? Or is there some other reason to split things up? Is there a compute cost to putting them elsewhere to consider as well?

joshuahannan commented:
We originally were going to completely remove type specifications in interfaces from the language, which would mean there would be no contract interface at all here and everything would have been contained in the `NonFungibleToken` contract, but we realized that it is nice to be able to specify the events and such, but that was after we had already put everything in a contract, so now I want to put everything back in the interface. 

We still want to have general utility functions and types for token such as `verifyCollectionTypes()` and scoped providers though, and those can't be in the standard interface because we can't write implementations of those functions in there, so we'd have to have them in a separate utility contract.

turbolent commented on contracts/NonFungibleToken-v2-ContractInterface.cdc:
```diff
@@ -0,0 +1,51 @@
+
+pub contract interface NonFungibleTokenInterface {
+
+    /// Event that is emitted when a token is withdrawn,
```
Events seem to be the last place in the standard where nested type requirements are used, e.g. there is no `Vault` anymore, unless I missed something.

Is enforcing that implementations of this interface provide the given event a requirement for the standard?
If not, it would allow us to remove the nested type requirement feature, which the Cadence team is eager to do for multiple reasons. Maybe we can find a different solution? What exactly should the standard ideally enforce / what functionality should it provide in terms of events?

joshuahannan replied:
I think most of us want the standard to enforce that tokens have standardized events that are emitted in the correct methods. Would there be another way to do that that doesn't use nested type requirements?

joshuahannan replied:
As we all know, events are basically the most important thing for apps to use to monitor what is happening on chain, so any standardization we can have there is good

turbolent replied:
> [...] most of us want the standard to enforce that tokens have standardized events that are emitted in the correct methods. 

Agreed. 

Is there a need to refer to the event types of the *interface* (`NonFungibleToken.Withdraw`)?

We currently don't have any means to enforce that a method emits an event, so maybe we should discuss that

turbolent commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,368 @@
+
+pub contract ExampleNFT: NonFungibleToken {
+
+    /// Standard events from the NonFungibleTokenInterface
```
```suggestion
    /// Standard events from the NonFungibleToken interface
```

psiemens reviewed (commented):
Great job drafting this up @joshuahannan!

I left some more general comments on the forum post, but LMK if it's better to discuss on this PR going forward: https://forum.onflow.org/t/streamlined-token-standards-proposal/3075/45?u=pete

psiemens commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,208 @@
+    ///
+    pub resource interface Provider {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun withdraw(withdrawID: UInt64): @AnyResource{NFT} {
```
That makes sense! I think it's worth it to keep it as `withdrawID` then.

bjartek commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,368 @@
+
+    pub event Withdraw(id: UInt64, from: Address?, type: Type)
+    pub event Deposit(id: UInt64, to: Address?, type: Type)
+    pub event Transfer(id: UInt64, from: Address?, to: Address?, type: Type)
```
Now that NFTs all implement resolveView is it worth it to add some of the views to this event _if_ they are present? Just things like Display and Edition/Serial will give so much value to make an event independent and human readable.

joshuahannan replied:
I think I'm ok with that. I'll add it to the proposal and we'll see what people think

bjartek commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,368 @@
+        }
+
+        /// Borrow the view resolver for the specified NFT ID
+        pub fun borrowViewResolver(id: UInt64): &AnyResource{MetadataViews.Resolver} {
```
This needs to return optional too I think.

joshuahannan replied:
That would require an upgrade to the metadataviews contract, which I guess is probably ok

joshuahannan commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,368 @@
+
+        /// Return the default public path for the collection
+        pub fun getDefaultPublicPath(): PublicPath {
+            return /public/cadenceExampleNFTCollection
```
how could we default implement that? We'd need access to the implementing contract to get its address, correct? afaik, from the collection public interface, we'd only have access to the address of the `NonFungibleToken` contract and the contract where the resource is stored. Maybe I'm missing something though

bjartek replied:
Yeah maybe I am just wishing for something I cannot have. I think this needs to be carefully considered up against other initiatiatives that are looking at removing paths.

joshuahannan commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,368 @@
+        switch nftType {
+            case Type<@ExampleNFT.NFT>():
+                return MetadataViews.NFTCollectionData(
+                    storagePath: /storage/cadenceExampleNFTCollection,
```
Probably not. I'll need to create a collection resource here and get the paths from there

austinkline replied:
Are we able to use `getCollectionData` for that? just feed the type of the nft itself in and it should work, right?

bjartek commented:
Since we do not have a universal collection anymore I honestly do not see the value of beeing able to have multiple NFTs in the same contract. It introduces changes that provide very little value IMHO.

If we could disconnect collections from NFTs it would be a totally different thing for me. But I do not see a concrete example of that beeing possible here. Can i create an NFT in a file that does not have a collection and just add that to another collection?

Also having collections with multiple different nft types is not trivial IMHO. So for me the benefit does not justify the cost here.

hier01 commented:
`        pub fun getAcceptedTypes(): {Type: Bool} {
            return {
                Type<@ExampleNFT.NFT>(): true
            }
        }`
Is there a reason not to hold the accepted types in an array in the Collection, instead of coded into this function?  That way an append mechanism could allow new types to be added.  Or deleted, but that would depend on whether the collection actually held any of the type being deleted.

joshuahannan commented:
@hier01 We're trying to make it so we can upgrade the existing standard to this one, and adding fields to existing types is not supported in upgrades, so we can't require that. If a project is still deploying a completely new NFT or collection, they are welcome to include that field in their collection though, even though it isn't required by the standard

Peixer commented:
@joshuahannan I was talking with @bjartek about a FLIP idea I'm working on, I think it'll be useful share it here with all you. 
@hichana @gpfurlaneto and me were discussing about how solve the [path conflicts](https://discord.com/channels/613813861610684416/1070200811973513236/1070203210096193576), the first idea we had was implementing a path registry like [this](https://gist.github.com/Peixer/705761ada3ec936a370b512dc9121a39) where all new smart contract would registry the path into the init function, example image below.  Contract type would be concat with the path name creating an unique path for each new smart contract.

It will create a centralized resource to handle all paths, using it to register new paths and consume paths. It can be used to improve the developer experience by a vscode plugin showing if the path is available even before deployment.

Please let me know your thoughts about it :) 

![image](https://user-images.githubusercontent.com/11881839/217142538-a96dade4-b579-4e7a-b1a6-daf401a14d8b.png)

joshuahannan commented:
@Peixer I think that standardization for paths is a great idea. I like the idea of your registry and correct me if I am wrong, but it looks like yours would only be for NFT contracts since it has the collection path paradigm built in? I think it would be better to make it more generalized. Can that be part of your FLIP?

Peixer commented:
@joshuahannan yeah, initially we thought about having this feature for collections. Not sure if I understood your point, make it more generalized would be a static function where anyone can create new paths using the standard? 

In the first draft we added a dictionary using the NFT contract type as a key, we would need to think about how to cover this part to be more generalized. Or maybe a contract registry could be made where the focus is on centralized contract information. How do you feel about this last part?

bluesign commented:
how much will be to register a path ?

bluesign commented:
Why registry I didn't understand that that part? 

Now we say path P is:  `P = Concat(contractType, path name)`; `contractType` we have access from the object, path name object can expose itself (via metadata, or simple standard interface) 

As we add contractType already, I don't think we need registry. A central registry would be a single point of failure, meaning that if it is compromised, the entire system could be at risk.

joshuahannan commented:
I talked with @Peixer and he is going to make a FLIP for his paths proposal so we can concentrate discussion for that there and keep this focused on the standards

bjartek commented on contracts/BasicNFT-v2.cdc:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        /// Gets the ID of the NFT, which here is the UUID
+        pub fun getID(): UInt64 { return self.uuid }
```
Can this be defaultly implemented in the NFT standard?

joshuahannan replied:
yes, good catch

bjartek commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,425 @@
+                case Type<MetadataViews.Editions>():
+                    // There is no max number of NFTs that can be minted from this contract
+                    // so the max edition field value is set to nil
+                    let editionInfo = MetadataViews.Edition(name: "Example NFT Edition", number: self.id, max: nil)
```
Can we use something else then id here? It is misleading to have uuid as edition

joshuahannan replied:
What would we use? We don't really have anything else in this contract, right? Should we just remove the edition view?

bjartek replied:
Either just have a counter that runs if they are all with the same name, or just leave it out TBH.

bjartek commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,425 @@
+        pub fun deposit(token: @AnyResource{NonFungibleToken.NFT}) {
+            let token <- token as! @ExampleNFT.NFT
+
+            emit Deposit(id: token.id, to: self.owner?.address, type: token.getType(),
```
Events emitted in interface?

joshuahannan replied:
We're gonna discuss that in the breakout session today. We need to make a FLIP for that, but I need help because I'm not sure what we would put in the FLIP

bjartek commented on contracts/ExampleNFT-v2.cdc:
```diff
@@ -0,0 +1,425 @@
+        self.account.save(<-collection, to: defaultStoragePath)
+
+        // create a public capability for the collection
+        self.account.link<&ExampleNFT.Collection{NonFungibleToken.CollectionPublic, MetadataViews.ResolverCollection}>(
```
This should be linked with all restrictions? Are the other ones gone?

joshuahannan replied:
yep, forgot to add `NonFungibleToken.Receiver` there

bjartek commented on contracts/MetadataViews.cdc:
```diff
@@ -25,7 +25,7 @@ pub contract MetadataViews {
     ///
     pub resource interface ResolverCollection {
-        pub fun borrowViewResolver(id: UInt64): &{Resolver}
+        pub fun borrowViewResolver(id: UInt64): &{Resolver}?
```
This is a really braking change. But i guess it is needed.

joshuahannan replied:
Everything here is going to be a breaking change as it is currently designed, so this won't make much of a difference

bjartek commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,257 @@
+    ///
+    /// The event that should be emitted when tokens are transferred from one account to another
+    ///
+    pub event Transfer(id: UInt64, from: Address?, to: Address?, type: Type,
```
I think a disucssion should happen on what views should be here. If we have an editioned nft having a single edition here makes tons of sense.

bjartek replied:
Also is returning type here and not identifier of thpe right? There was issues with serializing type right before iirc.

joshuahannan replied:
Added to the agenda for the breakout.

bjartek commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,257 @@
+    ///
+    pub resource interface NFT: MetadataViews.Resolver {
+        /// The unique ID that each NFT has
+        pub fun getID(): UInt64
```
I think this shoukd return uuid if that is possible.

joshuahannan replied:
added

bjartek commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,257 @@
+
+*/
+
+import MetadataViews from "./MetadataViews.cdc"
```
I think MetadataViews imports NonFungibleToken so this will not work?

joshuahannan replied:
yeah we're going to discuss that too

bluesign commented:
> For now, I also split the NFT metadata views out from the core metadata views contract so I could deploy everything and make sure it all works properly and typechecks. Ideally, we could actually have the metadata views split out like this, but idk if that will be possible.

I think extracting resolver interfaces are cleaner. With interfaces conforming to interfaces ( if we will have ) it will be backwards compatible (I guess)

joshuahannan commented:
Thanks for the feedback @bluesign! I did move the standard views back into `MetadataViews` and moved the `Resolver` and `ResolverCollection` interfaces into the `ViewResolver` contract. 

I also just pushed some more changes to use events in interfaces and I changed the standard to just be a contract and not an interface. I'm curious to hear if anyone has any feedback! I have a few things I want to discuss at the next breakout, so I'll start planning an agenda and scheduling the next breakout

bjartek commented:
Have we spoken about an event when something is Burned? At the moment is basically impossible to distinguish a Withdraw that escrow the resource elsewhere and a burn if you do not know what other event to look for.

joshuahannan commented:
@bjartek yep, it is in both proposals already: https://github.com/onflow/flow-nft/pull/126/files#diff-0aa93a6098e86167b39c0a5202f7cbdc953e8bd7ee949b594c2d2695b67372a8R120-R125

joshuahannan commented:
Just want to leave a note here that it might not be a good idea to include [this default implementation](https://github.com/onflow/flow-nft/blob/3acffe2472a3d2792a05a166e8177af3c8176ab8/contracts/NonFungibleToken-v2.cdc#L112-L114 ) because existing projects who upgrade might forget to include it and if they do and are using a different ID system, they would start returning uuid instead which could cause some problems.

It is nice to have for new projects though because they wouldn't even have to implement it, reducing the number of lines of code they have to write, but that might not be worth the risk.

Anyone else have any thoughts? I'm not fully decided yet.

bjartek commented:
There are two things i would like to discuss to be added to the standard. 

1. an event that is emitted like the others when an NFT is updated without a movement event. For doodles we use DoodleUpdated. This is very useful for indexer that wants to make a cache of nfts offchain.

2. A way to get attached/sub-nfts in a standard way for any nft. Personally i think it is overkill to say that an NFT can have an entire new NFTCollection but that is an option. Another options is some simple methods that are defaulty implemented to empty that says what sub-nfts you have (type and ids) and another method to get an sub-nft using type and id. 

I know there is uuid but it is currently not emited in any old events and there are conflicts in them. Personally I think just these two methods would be enough.

crash13override commented:
I totally agree with @bjartek about the need for those events, so that the real state of an NFT can be tracked properly. A Flovatar wearing an epic accessory for example is much more worth than without it, and it's important that people have a way to keep track of any change on the NFT.

joshuahannan commented:
Good suggestions! For the first one, I think that would be nice, but since we don't have nested type requirements any more, we can't enforce that certain events are included and still can't enforce that they are emitted properly. How would you propose we implement that and enforce that people use it properly?

For the second suggestion, that sounds like a metadata view, which we already have [an open issue](https://github.com/onflow/flow-nft/issues/146) for. I'm not sure that it should be part of the standard.

For both, if you'd like to include them, please make a PR or a comment with your suggestion! Y'all know much better about what you need than I do, so it would save a lot of back and forth if y'all made the proposal first and we can iterate on that. We also are going to discuss the new standards again at the smart contract open house on tuesday next week if we want to discuss it there too, but it would really help if we had your concrete proposal to discuss instead of just an idea.

bjartek commented:
> For the second suggestion, that sounds like a metadata view, which we already have https://github.com/onflow/flow-nft/issues/146 for. I'm not sure that it should be part of the standard.

Part of this can not be a view, what can be a view is exposing that you have sub-items and what ids you have. However for the borrowing of an sub-nft it would be most flexible if there was a method defaultly implmented as empty that allowed you to `borrowSubResource(type:Type, id:UInt64)` or something like that. I will make PR.

bjartek commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,203 @@
+    ///
+    access(all) event Deposit(id: UInt64, uuid: UInt64, to: Address?, type: String)
+
+    access(self) view fun emitNFTDeposit(id: UInt64, uuid: UInt64, to: Address?, type: String): Bool
```
Should these method take a reference to the resource and just get the fields instead like id/uuid/type? that way it is not possible to lie about those i guess?

joshuahannan replied:
it isn't possible to lie about the arguments because they get their values from the function arguments of `deposit` and withdraw

bjartek replied:
But somebody else can call them with fake data? Granted they can do that with fake address too?

joshuahannan replied:
The `emit` functions are `access(self)` so only functions in the standard can call them

bjartek commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,203 @@
+    ///
+    /// The event that should be emitted when tokens are transferred from one account to another
+    ///
+    access(all) event Transfer(id: UInt64, uuid: UInt64, from: Address?, to: Address?, type: String)
```
These can just go away can they not the emit method and the emit method?

bjartek commented:
I added a PR targeted to this branch with the update event.

bjartek commented:
@sisyphusSmiling had a point in flow discord today about `how do we now initialize an NFT collection`. There is an pattern currently where you can borrow a NFT contract address and get its collectionData and then initialize. In the new v2 ExampleNFT the contract does implement ViewResolver that gives the same, but there is no _enforcement_ of this. And Personally i would very much like there to be enforcement of this. 

So I guess i then question the move from a contract interface to a contract. Are there going to be an practical use of having multiple NFTs in a contract? In my eyes I do not think that the gains are worth the cost here.

bjartek commented on contracts/NonFungibleToken-v2.cdc:
```diff
@@ -0,0 +1,191 @@
+        access(all) fun deposit(token: @{NFT})
+
+        /// getSupportedNFTTypes returns a list of NFT types that this receiver accepts
+        access(all) view fun getSupportedNFTTypes(): {Type: Bool}
```
Personally i do not see much reason for this and the one below it anymore. We have pretty much gone away from a collection storing more then 1 nft type have we not?

bjartek replied:
If anything this should just return a singular type IMHO.

joshuahannan replied:
This is how it is currently on mainnet. Receivers have no obligation to only accept a single type, even if collections typically only hold one type. If we made this only accept a single type, we wouldn't be able to support use cases like a NFT switchboard and other such uses.

bjartek commented on contracts/utility/NFTForwarding.cdc:
```diff
@@ -2,46 +2,47 @@
 
-    pub event ForwardedNFTDeposit(id: UInt64, from: Address?)
-    pub event NFTForwarderRecipientChanged(forwarder: Address?)
+    access(all) event ForwardedNFTDeposit(id: UInt64, from: Address?)
```
Can this event contain type and UUID fields like NFT have added now?

joshuahannan replied:
yep, I'll add

bjartek commented on contracts/NonFungibleToken.cdc:
```diff
@@ -41,162 +36,160 @@ Collection to complete the transfer.
+                result.getLength() == 0: "The created collection must be empty!"
+            }
         }
 
```
```suggestion
/// Return a dictionary of all subNFTS if any
access(all) view fun getAvailableSubNFTS(): { Type: UInt64} {
    return {}
}
```

bjartek replied:
This can be done in a view, but since the other method is here I think this makes it more complete.

bjartek commented on contracts/NonFungibleToken.cdc:
```diff
@@ -41,162 +36,160 @@ Collection to complete the transfer.
-        pub var ownedNFTs: @{UInt64: NFT}
+        /// Return the NFT CollectionData View
+        /// has to be AnyStruct and cast to the view later to avoid circular dependency issues
+        access(all) fun getNFTCollectionDataView(): AnyStruct
```
Personally I do not see this as needed, we have lots of code to resolve this view other places. This is just extra code that everybody have to write.

joshuahannan replied:
I would agree if I trusted everyone to implement metadata views properly, but this forces everyone to implement this function and think about how to implement the collection data view instead of just skipping or returning the bare minimum, so i think it is pretty useful

bjartek replied:
But it returns AnyStruct? you can implement it with returning "bob". If it said you have to return NFTCollectionDataView it is entirely different.

joshuahannan replied:
the point is that it forces people to think about it, which IMO is worthwhile. I'll need to update the comment on the function to include more information and maybe a link to the guide, but I want people to be forced to think about how to do it and to look for resources

bjartek replied:
I get that, but why can it then not return NFTCollectionData? that way it is also far easier to use since you don`t have to cast and stuff.

joshuahannan replied:
Agreed, but that creates a circular dependency in `NonFungibleToken` because we would have to import `MetadataViews` and we already import `NonFungibleToken` in `MetadataViews` and I don't know how to make that work. If you know a way to make that work, let me know and I can update it

joshuahannan commented:
To follow up for anyone watching, we decided to change both standards back to interfaces. Summary of the discussion can be found in [the FLIP](https://github.com/onflow/flips/pull/56#issuecomment-1898982600)

sisyphusSmiling commented on contracts/NonFungibleToken.cdc:
```diff
@@ -41,162 +36,195 @@ Collection to complete the transfer.
-        pub fun getViews(): [Type] {
-            return []
+        /// unique ID for the NFT
+        access(all) let id: UInt64
```
I feel like I know the answer here, but I have to ask - with the introduction of Flow EVM, the ERC721 standard has me wondering if there's any desire to update `id` from `UInt64` to `UInt256`? Selfishly asking as it will otherwise be a potential pain point for bridged assets, not that changing it at this point wouldn't have it's difficulty & effort as it would likely require a migration.

joshuahannan replied:
I have no `attachment` to UInt64 for the id type, so if it is possible in a migration, then I'm okay with it. @turbolent What do you think? would it be possible to migrate all ID fields of NFTs to `UInt256`? We would potentially need to migrate uuids also, but I'm not sure.

@sisyphusSmiling It would also be good to understand how much of a headache it is for you to have this mismatch. If it is something you can handle relatively easily, then it might not be worth the hassle for trying to figure out a way to migrate safely if that is going to be difficult.

sisyphusSmiling replied:
I'm trying to avoid pigeonholing IDs for EVM-native NFTs bridging to Flow. The problem fundamentally comes down to the fact that ERC721 ids are `uint256`, meaning we have 1/4 available ids in Flow than might exist in EVM.

This presents two problems:
1. Accurately representing the NFT ID on Flow according to the NFT standard without collisions
2. Ensuring IDs that bridge to Flow are bridged back to their correct corresponding NFT

The only way I can think of to handle this is by having a `mod` field on bridged EVM-native NFTs. So imagine an ERC721 with id = UInt64.max + 1 would look like:

```cadence
access(all) resource NFT {
    access(all) let id: UInt64 = 1
    access(all) let mod: UInt8 = 1
}
```

where id = ERC721.id % UInt64.max and mod = ERC721.id / UInt64.max. I'm effectively wrapping the UInt64 based on the UInt256 value.

This solves 2 as I can recalculate the original `uint256` value from these two fields. However, I'm still left with the problem of 1 - collisions on IDs once NFTs are bridged to Flow are unavoidable.

austinkline replied:
What backend systems even support UInt256? A change like that could cause havoc in lots of places, does it makes sense to move this to a FLIP instead of here in a PR, or to schedule a discussion? It would have massive implications for every existing app that's already having to do a lot of work for 1.0

sisyphusSmiling replied:
Given the priority on finalizing the standards, I don't want to hold up progress. I raised it here as a point of discussion and to temp check sentiment on such a change, so thanks for adding a point of concern @austinkline. I don't doubt there would be impacts to codebases outside of the context of just this standard, but figured this is likely the only chance we'll have to make significant changes and this hasn't yet been discussed.

@joshuahannan based on what we have to discuss on Thursday, would you recommend raising this in our sync then? Meanwhile, happy to discuss async as I know the timeline for v2 standard delivery is compressed and high priority.

austinkline replied:
@sisyphusSmiling I think part of this will depend on how you envision an NFT being represented across chains here, because you could always let an EVM-origin nft use its resource uuid as the ID, and then maintain an EVM id field that has the proper number

sisyphusSmiling replied:
Great point @austinkline! Adding a field mapping to the EVM-defined ID makes sense in a world with different id types. And you're right, much less work for all involved to create new guidance moving forward than refactor existing code and offchain systems

joshuahannan commented:
We finalized the design of the standards at our last working group meeting and this PR has been updated with the intended final implementations of `NonFungibleToken`, `ViewResolver`, and `MetadataViews`. 

Anyone who is interested is welcome to take a look and leave any feedback or questions they want, but any fundamental changes to the standards will likely not be actioned unless the need is very strong because developers need to have a stable set of standards to work on their Cadence 1.0 upgrades with.

The other contracts are all either examples or utility contracts, so those designs aren't finalized yet. The tests and transactions also have not fully been updated yet, but will be in the coming weeks.

Thank you to everyone who contributed to the designs! We couldn't have done it without you. 😃

joshuahannan commented:
I just marked this PR as ready for review! There are still some tests that need to be added, but I consider most of the code to be ready! I'm welcome to hear any feedback about any of the choices in the example contracts and transactions so we can make sure we can set a good example for Cadence 1.0!

turbolent commented:
@joshuahannan Awesome! Maybe the PR title can be updated now? Is flow-go already updated to this?

joshuahannan commented:
@turbolent Good suggestion! I updated the PR titles. This was also included in my PR that got merged to the flow-go feature branch. Same with the latest flow-core-contracts

joshuahannan commented:
Looking for reviews on the standard code changes here if anyone has time! I'd like to get this merged soon! 😃

austinkline commented on contracts/MetadataViews.cdc:
```diff
@@ -377,162 +311,325 @@ pub contract MetadataViews {
+    ///         keys that are not wanted to become `Traits`
+    /// @return The generated Traits view
+    ///
+    access(all) fun dictToTraits(dict: {String: AnyStruct}, excludedNames: [String]?): Traits {
```
Should this be a view function?

austinkline commented on contracts/MetadataViews.cdc:
```diff
@@ -377,162 +311,325 @@ pub contract MetadataViews {
+    ///         keys that are not wanted to become `Traits`
+    /// @return The generated Traits view
+    ///
+    access(all) fun dictToTraits(dict: {String: AnyStruct}, excludedNames: [String]?): Traits {
```
This function isn't going to work anymore from where it is used most (`getViews`), we will need to rework it a bit looks like

joshuahannan replied:
This isn't used from `getViews`, is it? Isn't it used from `resolveView`?

loic1 commented on contracts/MetadataViews.cdc:
```diff
@@ -11,113 +12,38 @@ import NonFungibleToken from "NonFungibleToken"
-    /// resource (e.g. an NFT) can implement this interface to provide access to
-    /// the views that it supports.
-    ///
-    pub resource interface Resolver {
```
Do we need to add back `access(all) resource interface Resolver {}` and `access(all) resource interface ResolverCollection {}` for backwards compatibility with existing resources that conform to them?

turbolent replied:
AFAIK, those got removed here because they moved to `ViewResolver`

joshuahannan replied:
yep, and we already have the upgrade exception planned for the migration

nvdtf commented on README.md:
```diff
@@ -245,13 +192,18 @@ but without most of the downsides.
 
-1. Deploy `NonFungibleToken.cdc`
-2. Deploy `ExampleNFT.cdc`, importing `NonFungibleToken` from the address you deployed it to.
+1. Deploy `ViewResolver.cdc`
```
might be helpful to mention that emulator deploys the standards to the service account on startup.

nvdtf commented on contracts/MetadataViews.cdc:
```diff
@@ -11,113 +12,38 @@ import NonFungibleToken from "NonFungibleToken"
+        access(all) let thumbnail: {File}
 
-        init(
+        view init(
```
Curious how come this is `view` but also changing state values? I see this pattern in other places, so for sure it's something I don't know about.

joshuahannan replied:
It can be `view` because it isn't changing any existing state. It is a newly created struct, so it is basically treated like a getter function

joshuahannan replied:
I agree it isn't super intuitive at a first glance though

nvdtf commented on contracts/MetadataViews.cdc:
```diff
@@ -200,125 +126,133 @@ pub contract MetadataViews {
     ///
     /// @param viewResolver: A reference to the resolver resource
-    /// @return An optional Serial struct
+    /// @return A optional License struct
```
```suggestion
    /// @return An optional License struct
```

nvdtf commented on contracts/MetadataViews.cdc:
```diff
@@ -200,125 +126,133 @@ pub contract MetadataViews {
+    /// Helper to get ExternalURL in a typesafe way
+    ///
+    /// @param viewResolver: A reference to the resolver resource
+    /// @return A optional ExternalURL struct
```
```suggestion
    /// @return An optional ExternalURL struct
```

nvdtf commented on scripts/get_collection_ids.cdc:
```diff
@@ -3,13 +3,12 @@
-        .getCapability(collectionPublicPath)
-        .borrow<&{NonFungibleToken.CollectionPublic}>()
-        ?? panic("Could not borrow capability from public collection at specified path")
+    let collectionRef = account.capabilities.borrow<&{NonFungibleToken.Collection}>(
```
When testing migration, I had to change the above borrow to this:
```
account.capabilities.borrow<&{
        NonFungibleToken.CollectionPublic,
        ExampleNFT.ExampleNFTCollectionPublic,
        ViewResolver.ResolverCollection
    }>(
        /public/exampleNFTCollection
      ) ?? panic("Could not borrow capability from collection at specified path")
```
Is this something only after migration?

loic1 replied:
@SupunS do we maybe need to add the same special case here as you did for interface conformance so that we can use the `&{NonFungibleToken.Collection}` restricted type? 👀

SupunS replied:
This is actually not related to the contract update restrictions (https://cadence-lang.org/docs/1.0/language/contract-updatability), to which the special casing was done. Those restrictions are applicable to field types and conformances only, just to prevent existing data from getting corrupted.

The type here, can be changed freely. However, the catch is, the stored value in that path also gets migrated during the storage migration (There is some detailed explanation [here](https://github.com/onflow/flips/blob/main/cadence/20230601-entitlements-migration.md) as to why the values are migrated this way). So even if you use the old type, when the script is executed, it will complain saying the stored value doesn't match the requested type.

_Just for some additional context: The migration has two steps: 1) Updating the staged contracts / contract code. 2) Migrating existing data. This particular problem is related to the step (2), where as, the type-validations/contract-update-restrictions are applicable only to the step (1)._

nvdtf replied:
It sounds like with the migration process, there would some differences with how stored values can be accessed that might not apply to contracts starting from 1.0 as shown in this example.
Is there a paragraph/doc in the migration guide capturing this information?

SupunS replied:
>  It sounds like with the migration process, there would some differences with how stored values can be accessed

Yes. Don't think we have any documentation around that yet. Good idea to add a section explaining this. Could you please create an issue in cadence repo?

nvdtf replied:
Good idea, just opened https://github.com/onflow/cadence/issues/3206
Feel free to edit or add any comments for clarification.

bluesign replied:
shouldn't it be:
```cadence 
account.capabilities.borrow<&{
        NonFungibleToken.CollectionPublic
    }>(
        /public/exampleNFTCollection
      ) ?? panic("Could not borrow capability from collection at specified path")
```

SupunS replied:
I actually was wondering the same. @dsainati1 do you know whether this is expected?

SupunS replied:
ah ok, so it seems the original capability was linked with all thee interfaces: https://github.com/onflow/flow-nft/blob/4c0c684baed98e2ffd7a8f9d27973c1db5dd1239/contracts/ExampleNFT.cdc#L398

So when the link gets converted to a capability, it retains the same intersection type.

joshuahannan replied:
That might be because the way it is linked before the migration doesn't get migrated cleanly. Because `{NonFungibleToken.Collection}` didn't exist before, I guess that the migration doesn't pick up that is is now possible to borrow as that type. The way I changed it to here is the way that I would prefer people do it in the future though. Not sure if that is more or less important than making it clearer for migrations

bluesign replied:
@SupunS 
> So when the link gets converted to a capability, it retains the same set of interfaces.
but doesn't it mean any one of them would work?

SupunS replied:
ah yes, it should (and does) work with `&{NonFungibleToken.CollectionPublic}`. I got it confused with the `NonFungibleToken.Collection` and didn't realize these two are two different interfaces (thought one got replaced with the other, but they both exist)

nvdtf reviewed (commented):
I had an initial review and had some questions. The general state of changes looks good so far to me.

nvdtf commented on transactions/mint_nft.cdc:
```diff
@@ -46,41 +44,34 @@ transaction(
-            count = count + 1
-        }
-
+        // // Create the royalty details
```
Should we add a comment on how to handle royalties in NFTv2?
Commented code might suggest its not supported any more

joshuahannan replied:
I commented it out because I was testing other things and forgot to uncomment it. I'll add something back for royalties

joshuahannan commented on transactions/test/upgrade_nft_contract.cdc:
```diff
@@ -0,0 +1,8 @@
+
+transaction(code: [UInt8]) {
```
Eventually, but I don't think it is a high priority right now

joshuahannan commented:
@nvdtf I addressed all your review comments and made almost all of the changes you suggested. Let me know if you have any other comments! Thank you so much for the review! :)

cody-evaluate commented:
With `Collection.ownedNFTs` no longer being standard (or really even accessible), there needs to be an efficient and standard way to loop through all owned NFT IDs in a collection. Calling `Collection.getIDs()` has increasingly common compute limit issues for large collections. 

We currently rely on `collection.ownedNFTs.forEachKey` heavily in our production scripts to iterate through all NFT IDs in a collection in an efficient way.

My suggestion is to add the following to the `CollectionPublic` standard interface:

`access(all) fun forEachID(_ f: fun (UInt64): Bool): Void`

implementation would look like this for legacy contracts (or really however the developer sees fit)

```
access(all) resource Collection: NonFungibleToken.Collection {

      access(all) fun forEachID(_ f: fun (UInt64): Bool): Void {
      
          self.ownedNFTs.forEachKey(f)
      }

      ...
}
```

joshuahannan commented:
Thanks for the comment @cody-evaluate! I'll update the `getLength` implementation in `ExampleNFT`.

I agree that what you are asking for would be very nice to have and if it wasn't so late in the process, I would put it in with no hesitation so that every project would have to implement it. As I said in Discord, we've promised to the community that we won't introduce any more breaking changes to the standard, which this would be because it would require that people implement it. We could put it in there with a default implementation though, which would not break any contracts that have already been updated. The downside though is that since there is a default implementation, contracts don't have to implement the function, meaning that there is no guarantee that every NFT has a proper implementation of this method. Is that okay?

cody-evaluate commented:
@joshuahannan its better than nothing i guess but this should really be required and even though its late, we are still in the staging period so theres really no other time to do something like this. The biggest issue is this removes a necessary piece of functionality that has existed in production since the beginning with no replacement currently.

joshuahannan commented:
I understand, but we had no idea that anyone was actually using `ownedNFTs` for anything, so removing it did not seem like any problem to us and we didn't hear any feedback about this in the year+ that we were working on the new standard until now. I'll see what I can do

cody-evaluate commented:
this would be an example use case in a script

```
import NonFungibleToken from 0x1d7e57aa55817448

pub fun main(ownerAddress: Address): [&{NonFungibleToken.INFT}] {

    let limit = 10

    let response: [&{NonFungibleToken.INFT}] = []

    let account = getAuthAccount(ownerAddress)

    account.forEachStored(fun (path: StoragePath, type: Type): Bool {

        if (!type.isSubtype(of: Type<@NonFungibleToken.Collection>())) {

            return true
        }

        let storageCollection = account.borrow<&NonFungibleToken.Collection>(from: path)!

        storageCollection.ownedNFTs.forEachKey(fun (nftId: UInt64): Bool {

            let nft = storageCollection.borrowNFT(id: nftId)

            response.append(nft)

            return response.length < limit
        })

        return response.length < limit
    })

    return response
}
```

cody-evaluate commented:
here is a working example of getting the NFT ID at index `150000` on a massive collection

// you cant even get index `0` in this example because arrays are value types in cadence
```
import NonFungibleToken from 0x1d7e57aa55817448

pub fun main(): UInt64 {

	let index = 150000

	let account = getAuthAccount(0x1e3c78c6d580273b)

	let storageCollection = account.borrow<&NonFungibleToken.Collection>(from: /storage/LNVCTCollection)!

	// compute limit
	return storageCollection.getIDs()[index]
}
```

this works for the same collection

```
import NonFungibleToken from 0x1d7e57aa55817448

pub fun main(): UInt64? {

	let limit = 150000

	let account = getAuthAccount(0x1e3c78c6d580273b)

	var index = 0

	var response: UInt64? = nil

	let storageCollection = account.borrow<&NonFungibleToken.Collection>(from: /storage/LNVCTCollection)!

	storageCollection.ownedNFTs.forEachKey(fun (nftId: UInt64): Bool {

		response = nftId

		index = index + 1

		return index < limit
	})

	return response
}
```

joshuahannan commented:
I think that it is time to merge this to master! If nobody has any issues with me merging this, I'll merge it on Monday next week. I'm going to create a branch that is based off current master so that we still have a record of the cadence 0.42 versions of the contracts.

turbolent commented:
Amazing work everyone! 👏
